Rubberduck.VBEEditor/Class1.cs

```diff
@@ -0,0 +1,964 @@
+        IntPtr Object1;
+        IntPtr Object2;
+        public IntPtr TypeLib;
+        IntPtr Placeholder1;
+        IntPtr Placeholder2;
+        IntPtr RefCount;
+    }
+
+    // A ITypeLib object hosted by the VBE, also providing Prev/Next pointers for a double linked list of all loaded project ITypeLibs
+    [StructLayout(LayoutKind.Sequential)]
+    struct TypeLibObj_VBE
+    {
+        IntPtr vTable1;     // ITypeLib vtable
+        IntPtr vTable2;
+        IntPtr vTable3;
+        public IntPtr Prev;
+        public IntPtr Next;
+    }
+
+    // IVBProjectEx_VBE, obtainable from a VBE hosted ITypeLib in order to access a few extra features...
+    [ComImport(), Guid("DDD557E0-D96F-11CD-9570-00AA0051E5D4")]
+    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    interface IVBProjectEx_VBE
+    {
+        void Placeholder1();
+        void Placeholder2();
+        int VBE_LCID();
+        void Placeholder3();
+        void Placeholder4();
+        void Placeholder5();
+        void Placeholder6();
+        void Placeholder7();
+        string get_ConditionalCompilationArgs();
+        void set_ConditionalCompilationArgs(string args);
+    }
+
+    // AggregateSingleInterface is used to ensure that a wrapped COM object only responds to a specific interface
+    // In particular, we don't want them to respond to IProvideClassInfo, which is broken in the VBE for some ITypeInfo implementations 
+    public class AggregateSingleInterface<T> : ICustomQueryInterface, IDisposable
+        where T : class
+    {
+        private IntPtr _outerObject;
+
+        public AggregateSingleInterface(IntPtr outerObject)
+        {
+            _outerObject = outerObject;
+            Marshal.AddRef(_outerObject);
+        }
+
+        private bool _isDisposed;
+        public void Dispose()
+        {
+            if (_isDisposed) return;
+            _isDisposed = true;
+            Marshal.Release(_outerObject);
+        }
+
+        public CustomQueryInterfaceResult GetInterface(ref Guid iid, out IntPtr ppv)
+        {
+            ppv = IntPtr.Zero;
+            if ((!_isDisposed) && (iid == typeof(T).GUID))       // no need to offer IID_IUnknown here, as it is handled by the aggregation object
+            {
+                ppv = _outerObject;
+                Marshal.AddRef(_outerObject);
+                return CustomQueryInterfaceResult.Handled;
+            }
+            return CustomQueryInterfaceResult.Failed;
+        }
+    }
+
+    [ComImport(), Guid("00020400-0000-0000-C000-000000000046")]
+    [InterfaceType(ComInterfaceType.InterfaceIsIDispatch)]
+    public interface IDispatch
+    {
+    }
+
+    // A compatible version of ITypeInfo, where COM objects are outputted as IntPtrs instead of objects
+    [ComImport(), Guid("00020401-0000-0000-C000-000000000046")]
+    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
+    public interface ITypeInfo_VBE
+    {
+        void GetTypeAttr(out IntPtr ppTypeAttr);
```

Comment:
Do we need to have the `[PreserveSig]` attributes?

Reply:
no, it's easier to leave it off, as otherwise we have to deal with all the HRESULTs ourselves in the wrappers